programs/mango-v4/src/instructions/perp_liq_with_token.rs

```diff
@@ -0,0 +1,365 @@
+            .min(-liqee_settleable_pnl)
+            .max(I80F48::ZERO);
+        settlement = max_settlement_liqee
+            .min(I80F48::from(max_liab_transfer))
+            .max(I80F48::ZERO);
+        if settlement > 0 {
+            liqor_perp_position.record_liquidation_quote_change(-settlement);
+            liqee_perp_position.record_settle(-settlement);
+
+            // Update the accounts' perp_spot_transfer statistics.
+            let settlement_i64 = settlement.round_to_zero().checked_to_num::<i64>().unwrap();
+            cm!(liqor_perp_position.perp_spot_transfers += settlement_i64);
+            cm!(liqee_perp_position.perp_spot_transfers -= settlement_i64);
+            cm!(liqor.fixed.perp_spot_transfers += settlement_i64);
+            cm!(liqee.fixed.perp_spot_transfers -= settlement_i64);
+
+            let now_ts: u64 = Clock::get()?.unix_timestamp.try_into().unwrap();
+
+            // Transfer token balance
+            let liqor_token_position = liqor.token_position_mut(settle_token_index)?.0;
+            let liqee_token_position = liqee.token_position_mut(settle_token_index)?.0;
+            settle_bank.deposit(liqor_token_position, settlement, now_ts)?;
+            settle_bank.withdraw_without_fee(
+                liqee_token_position,
+                settlement,
+                now_ts,
+                oracle_price,
+            )?;
+            liqee_health_cache.adjust_token_balance(&settle_bank, -settlement)?;
+
+            emit!(PerpLiqWithTokenLog {
+                mango_group,
+                liqee: ctx.accounts.liqee.key(),
+                liqor: ctx.accounts.liqor.key(),
+                perp_market_index: perp_market_index,
+                settlement: settlement.to_bits(),
+            });
+
+            msg!("liquidated pnl = {}", settlement);
+        }
+    };
+    let max_liab_transfer = cm!(I80F48::from(max_liab_transfer) - settlement);
+
+    //
+    // Step 2: bankruptcy
+    //
+    // Remaining pnl that brings the account into negative init health is either:
+    // - taken by the liqor in exchange for spot from the insurance fund, or
+    // - wiped away and socialized among all perp participants
+    //
+    let insurance_transfer = if settlement == max_settlement_liqee {
+        let liqee_perp_position = liqee.perp_position_mut(perp_market_index)?;
+        let liqee_pnl = liqee_perp_position.pnl_for_price(&perp_market, oracle_price)?;
+
+        let max_liab_transfer_from_liqee = (-liqee_pnl).min(-liqee_init_health).max(I80F48::ZERO);
+        let liab_transfer = max_liab_transfer_from_liqee
+            .min(max_liab_transfer)
+            .max(I80F48::ZERO);
+
+        // Available insurance fund coverage
+        let insurance_vault_amount = if perp_market.elligible_for_group_insurance_fund() {
+            ctx.accounts.insurance_vault.amount
+        } else {
+            0
+        };
+
+        let liquidation_fee_factor = cm!(I80F48::ONE + perp_market.liquidation_fee);
```

Comment:
is liquidation_fee required, if insurance fund is usdc?

Reply:
It's the liqor incentive to make sure they'll want to call it. It may not be strictly needed - similarly to the settlement part - because liqors should be happy to take negative perp pnl and token assets: it doesn't come with risk and allows them to collect interest until it's settled.